src/serialize-closure.ts

```diff
@@ -0,0 +1,1419 @@
+          "const",
+          uniqueName(),
+          newExpr(idExpr("Proxy"), [target, handler])
+        );
+      }
+      throw new Error(
+        `cannot reverse Proxy - make sure you are compiling with Functionless`
+      );
+    } else if (typeof value === "object") {
+      if (Globals.has(value)) {
+        return emitVarDecl("const", uniqueName(), Globals.get(value)!());
+      }
+
+      if (options?.serialize) {
+        const result = options.serialize(value);
+        if (!result) {
+          // do not serialize
+          return emitVarDecl("const", uniqueName(), undefinedExpr());
+        } else if (value === true || typeof result === "object") {
+          value = result;
+        } else {
+          return serialize(result);
+        }
+      }
+
+      const mod = requireCache.get(value);
+
+      if (mod && options?.requireModules !== false) {
+        return serializeModule(value, mod);
+      }
+
+      if (
+        Object.hasOwn(value, "__defineGetter__") &&
+        value !== Object.prototype
+      ) {
+        // heuristic to detect an Object that looks like the Object.prototype but isn't
+        // we replace it with the actual Object.prototype since we don't know what to do
+        // with its native functions.
+        return serialize(Object.prototype);
+      }
+
+      if (
+        Object.hasOwn(value, "constructor") &&
+        typeof value.constructor === "function" &&
+        value.constructor.prototype === value
+      ) {
+        // this is the prototype value of a Function/class
+        const clazz = serialize(value.constructor);
+        return singleton(value, () => propAccessExpr(clazz, "prototype"));
+      }
+
+      const prototype = Object.getPrototypeOf(value);
+
+      // serialize the prototype first
+      // there should be no circular references between an object instance and its prototype
+      // if we need to handle circular references between an instance and prototype, then we can
+      // switch to a strategy of emitting an object and then calling Object.setPrototypeOf
+      const serializedPrototype =
+        prototype !== Object.prototype ? serialize(prototype) : undefined;
+
+      // emit an empty object with the correct prototype
+      // e.g. `var vObj = Object.create(vPrototype);`
+      const obj = singleton(value, () =>
+        emitVarDecl(
+          "const",
+          uniqueName(),
+          serializedPrototype
+            ? callExpr(propAccessExpr(idExpr("Object"), "create"), [
+                serializedPrototype,
+              ])
+            : objectExpr({})
+        )
+      );
+
+      defineProperties(value, obj);
+
+      return obj;
+    } else if (typeof value === "function") {
+      if (options?.serialize) {
+        const result = options.serialize(value);
+        if (result === false) {
+          // do not serialize
+          return undefinedExpr();
+        } else if (result !== true) {
+          value = result;
+        }
+      }
```

Comment:
IMO, the immutable version is actually more annoying. If the `result === true` I need to set the `updatedValue = value` since `true` indicates that the original value should be serialized.

I can put this logic in once place though! 👍

Reply:
It's not straightforward to generalize this contract.